Working bolton model without unit tests.

cchoquette opened the pull request:
Bolton model tested on binary classification. 
No testing or additional loss functions implemented yet (TBD). 
Requires tf2.0.0 or greater.

cchoquette committed:
Working bolton model without unit tests.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+# limitations under the License.
+"""Loss functions for bolton method"""
+
+from __future__ import absolute_import
```
Is the implementation compatible for both python 2 and 3?

cchoquette replied:
Tensorflow is and this implementation should be as well

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+  """
+  def __init__(self,
+               reg_lambda: float,
+               c: float,
```
I think using a capital C makes it more consistent with sklearn's tradition. Check the document for LinearSVC for the same parameter C. The docstring for this parameter can also be similar to the one used for LinearSVC.

jichaoz replied:
I think reg_lambda and c are more a concept for a concrete loss function. I understand that in the case of HuberLoss and CrossEntropyLoss, they both need these two parameters, but these might not apply to all potential strongly convex loss functions. I feel the original design in the mxnet implementation to have StronglyConvexLoss as an interface is conceptually more clean.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+    super(StrongConvexLoss, self).__init__(reduction=reduction,
+                                           name=name,
+                                           **kwargs)
+    self._sample_weight = tf.Variable(initial_value=c,
```
Why is this variable named _sample_weight?

cchoquette replied:
My mistake - was miss name

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+    self.dtype = dtype
+
+  def radius(self):
+    """Radius of R-Ball (value to normalize weights to after each batch)
```
Can you make the docstring for the 4 abstract function more clear? I remember the paper has some good definition for them. We can also add links to the paper for user to learn more about these parameters.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+    return tf.math.reduce_max(class_weight)
+
+
+class Huber(StrongConvexLoss, losses.Huber):
```
To differentiate our version of the loss function, lets name it StronglyConvexHuberLoss or StronglyConvexHuber.

jichaoz replied:
The loss function we implemented is different than any of those implemented in tensorflow. The paper has the formula.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,280 @@
+    Returns:
+      Loss values per sample.
+    """
+    return super(Huber, self).call(y_true, y_pred, **self._fn_kwargs) * \
```
Do we need to override this method? It seems we didn't make any change.

cchoquette replied:
As discussed in person, we need to to include C in the loss calculation

jichaoz reviewed (changes requested):
I havn't completed the whole review yet, but let's discuss the comments so far.

cchoquette committed:
Working bolton model without unit tests.
-- update to include pull request changes
changes include:
parameter renaming,
changing to mixin,
moving model to compile,
additional tests,
fixing huber loss

cchoquette committed:
Working bolton model without unit tests.
-- moving to Bolton Optimizer
Model is now just a convenient wrapper and example for users.
Optimizer holds ALL Bolton privacy requirements.
Optimizer is used as a context manager, and must be passed the model's layers.
Unit tests incomplete, committing for visibility into the design.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+    """Smoothness, beta
+
+    Args:
+      class_weight: the class weights used.
```
It will be great to describe the shape of class_weight.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+               radius_constant: float,
+               delta: float,
+               reduction: str = losses_utils.ReductionV2.SUM_OVER_BATCH_SIZE,
+               name: str = 'huber',
```
I feel we don't need this name param and can set it to our like.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+    """
+    # return super(StrongConvexHuber, self).call(y_true, y_pred) * self._sample_weight
+    h = self._fn_kwargs['delta']
+    z = y_pred * y_true
```
Assuming y_true is label encoded, the formula requires one-hot encode y_true with (-1 for negative and 1 for positive)

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+    four = tf.constant(4, dtype=self.dtype)
+
+    if z > one + h:
+      return z - z
```
why don't we simply return 0?

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+    return tf.math.reduce_max(class_weight)
+
+
+class StrongConvexHuber(losses.Huber, StrongConvexMixin):
```
We defined our own call(). Why do we still need to inherit from losses.Huber? Can we inherit from losses.Loss?

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+      this loss function to be strongly convex.
+    :return:
+    """
+    return L1L2(l2=self.reg_lambda)
```
In the paper, all the derivation of the formula are based on the assumption of the regularization term lambda/2 * ||W|| ^ 2. The L1L2 implementation of tensorflow is missing the 1/2 term. We might have to change the l2=self.reg_lambda/2.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+               C: float,
+               radius_constant: float,
+               from_logits: bool = True,
+               label_smoothing: float = 0,
```
I think we should not support this parameter because its impact to privacy guarantee is unknown.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -0,0 +1,588 @@
+      this loss function to be strongly convex.
+    :return:
+    """
+    return L1L2(l2=self.reg_lambda)
```
Similar to the case for strongly huber loss, we might need to set l2 to self.reg_lambda/2 due to the difference between the paper's formula and tensorflow's implementation

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+
+class Bolton(optimizer_v2.OptimizerV2):
+  """
+    Bolton optimizer wraps another tf optimizer to be used
```
The document seems a bit out-of-date. We should also cover how we design Bolton optimizer as context manager.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+      per_class_epsilon = self.epsilon / (output_dim)
+      l2_sensitivity = (2 *
+                        loss.lipchitz_constant(class_weight)) / \
+                       (loss.gamma() * data_size)
```
I remember the batch size is also in the denominator of the formula. Can you double check with Chang? You can also see https://gpartners.atlassian.net/wiki/spaces/IT/pages/540115020/Bolt-on.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+    """
+    loss = self.loss
+    distribution = self.noise_distribution.lower()
+    if distribution == _accepted_distributions[0]:  # laplace
```
Is there any other alternative choice for the noise distribution?

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+    """helper method to normalize the weights to the R-ball.
+
+    Args:
+        r: radius of "R-Ball". Scalar to normalize to.
```
r is not a parameter any more.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+
+    """
+    r = self.loss.radius()
+    for layer in self.layers:
```
This block of code feels not as concise as other code. It seems checking force before eager mode could make the code flowing more smoothly.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -0,0 +1,357 @@
+      raise ValueError('Detected noise distribution: {0} not one of: {1} valid'
+                       'distributions'.format(noise_distribution,
+                                              _accepted_distributions))
+    self.noise_distribution = noise_distribution
```
Why are we doing this?

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+
+    Args:
+        n_classes: number of output classes to predict.
+        epsilon: level of privacy guarantee
```
epsilon, noise_distribution, and weights_initializer are not accepted as parameters here any more.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+    """See super class. Default optimizer used in Bolton method is SGD.
+
+    """
+    for key, val in StrongConvexMixin.__dict__.items():
```
We should be able to enforce this using a more natural approach. We can declare the methods in StrongConvexMixin to be abstract method and then check the loss passed in is derived from StronglyConvexMixin.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+      # re-instantiate the model each time they want to change the penalty
+      # weighting
+      self._layers_instantiated = True
+    self.output_layer.kernel_regularizer.l2 = loss.reg_lambda
```
Do we still need this line?

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+
+  def compile(self,
+              optimizer='SGD',
+              loss=None,
```
We should always expect the user to provide the loss and thus not providing a default value.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+  def compile(self,
+              optimizer='SGD',
+              loss=None,
+              metrics=None,
```
Unless you are touching a parameter, I suggest you simply use **kwargs to forward them.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+          x=None,
+          y=None,
+          batch_size=None,
+          epochs=1,
```
Unless you are touching a parameter, I suggest you to use **kwargs to forward them.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -0,0 +1,551 @@
+    input_image, real_image = load(image_file)
+    input_image, real_image = resize(input_image, real_image,
+                                     IMG_HEIGHT, IMG_WIDTH)
+    input_image, real_image = normalize(input_image, real_image)
```
Does this step guarantee the ||x|| <= 1?

cchoquette replied:
Sorry, that accidentally got pushed. This was just my testing script to ensure everything worked.

jichaoz reviewed (changes requested):
It looks great overall and seems the optimizer as context manager approach should work. Please address the comments and submit review again.

cchoquette committed:
Bolton created as optimizer with context manager usage.
Unit tests included.
Additional loss functions TBD.

cchoquette committed:
Update Huber loss regularization term and some small changes across loss parameters.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -137,40 +137,43 @@ def __init__(self,
     super(StrongConvexHuber, self).__init__(
-        delta=delta,
-        name=name,
+        name='huber',
```
You might want to name this as 'StronglyConvexHuber' to be consistent and different than the one tf provides.

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -137,40 +137,43 @@ def __init__(self,
 
     if z > one + h:
-      return z - z
+      return _ops.convert_to_tensor_v2(0, dtype=self.dtype)
```
Is there any difference than using tf.constant(0)?

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -228,7 +231,6 @@ def __init__(self,
                from_logits: bool = True,
                label_smoothing: float = 0,
```
Why don't we simply remove this parameter since the impact is unknown?

jichaoz commented on privacy/bolton/loss.py:
```diff
@@ -250,11 +254,11 @@ def __init__(self,
     super(StrongConvexBinaryCrossentropy, self).__init__(
         reduction=reduction,
-        name=name,
+        name='binarycrossentropy',
```
We might better use 'stronglyconvexbinarycrossentropy' to be different than the one from tf.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -19,9 +19,74 @@
+                           )
+    dtype = self.beta.dtype
+    one = tf.constant(1, dtype)
+    return tf.math.minimum(tf.math.reduce_min(one/self.beta),
```
why do we need tf.math.reduce_min considering self.beta is a scalar?

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -31,11 +96,24 @@ class Bolton(optimizer_v2.OptimizerV2):
     passed, "Bolton" enables the bolton model to control the learning rate
     based on the strongly convex loss.
```
Shall we also mention 'project to R ball' and inject noise as part of the logic to provide privacy guarantee?

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -58,13 +137,19 @@ def __init__(self,
                                 'layers',
-                                '_model'
+                                'batch_size',
+                                '_is_init'
```
It seems _is_init is not used.

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -205,113 +262,110 @@ def __setattr__(self, key, value):
+                                  self.loss.gamma()
+                                  )
+    self.epsilon = _ops.convert_to_tensor_v2(epsilon, dtype=self.dtype)
+    self.class_weights = _ops.convert_to_tensor_v2(class_weights,
```
Is there a reason we cannot simply use tf.const() for these?

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -205,113 +262,110 @@ def __setattr__(self, key, value):
-    self.class_weights = class_weights
-    self.n_samples = n_samples
-    self.n_classes = n_classes
+    self.learning_rate.initialize(self.loss.beta(class_weights),
```
How is learning_rate limit applied?

jichaoz commented on privacy/bolton/optimizer.py:
```diff
@@ -328,30 +382,21 @@ def __exit__(self, *args):
     self.n_samples = None
     self.input_dim = None
-    self.n_classes = None
+    self.n_outputs = None
```
Where is self.n_outputs used? It seems we get output_dim from layer.units.

jichaoz commented on privacy/bolton/model.py:
```diff
@@ -41,41 +39,28 @@ class BoltonModel(Model):
     self._dtype = dtype
 
-  def call(self, inputs):
+  def call(self, inputs, training=False):  # pylint: disable=arguments-differ
```
How are we using the parameter 'training'?

cchoquette committed:
Bolton implementation and unit tests. Has two pre-implemented loss functions.

cchoquette committed:
Minor changes + tutorial

cchoquette committed:
Changes for pylint.

cchoquette committed:
Minor changes to function arguments
